w12h03/test/pgdp/shuttle/ShuttleProcessorTests.java

```diff
@@ -0,0 +1,181 @@
+            return super.toString().replace("\r", "");
+        }
+    };
+    private static final PrintStream stdOut = System.out;
+
+    @AfterEach
+    public void resetBuffer() {
+        out.reset();
+    }
+
+    @BeforeAll
+    public static void setUpStreams() {
+        System.setOut(new PrintStream(out));
+    }
+
+    @AfterAll
+    public static void restoreStreams() {
+        System.setOut(stdOut);
+    }
+
+    @Test
+    @DisplayName("Should evaluate priority task first")
+    public void testEvaluatePriority() throws InterruptedException, NoSuchFieldException, IllegalAccessException {
+        var taskchecker = new TaskChecker(null, null);
+        var sp = new ShuttleProcessor(taskchecker);
+        var taskGen = new TestTaskGenerator(0, 5, 0);
+
+        var task1 = taskGen.generateTask();
+        var task2 = taskGen.generateTask();
+        var task3 = taskGen.generateTask();
+        var task4 = taskGen.generateTask();
+
+        var taskList = List.of(task1, task3, task2, task4);
+
+        getPrioTaskQueue(sp).add(task1);
+        getTaskQueue(sp).add(task2);
+        getPrioTaskQueue(sp).add(task3);
+        getTaskQueue(sp).add(task4);
+
+        sp.start();
+        Thread.sleep(30);
+
+        var taskQueue = getTaskQueue(taskchecker);
+        for(int i = 0; i < 4; i++) {
+            assertEquals(taskList.get(i), taskQueue.poll());
+        }
+
+        sp.shutDown();
+        Thread.sleep(30);
+
+        assertEquals("ShuttleProcessor shutting down.\n", out.toString());
+    }
+
+
+    @Test
+    @DisplayName("Should still work when task are added while running")
+    public void testAddTaskWhileRunning() throws InterruptedException, NoSuchFieldException, IllegalAccessException {
+        var taskchecker = new TaskChecker(null, null);
+        var sp = new ShuttleProcessor(taskchecker);
+        var taskGen = new TestTaskGenerator(0, 5, 0);
+
+        sp.start();
+        Thread.sleep(10);
+
+        var task1 = taskGen.generateTask();
+        var task2 = taskGen.generateTask();
+        var task3 = taskGen.generateTask();
+        var task4 = taskGen.generateTask();
+
+        var taskList = List.of(task1, task3, task2, task4);
+
+        synchronized (sp) {
+            getPrioTaskQueue(sp).add(task1);
+            getTaskQueue(sp).add(task2);
+            sp.notify();
+            Thread.sleep(5);
+            getPrioTaskQueue(sp).add(task3);
+            getTaskQueue(sp).add(task4);
+            sp.notify();
+        }
+
+        Thread.sleep(30);
+
+        var taskQueue = ReflectionHelper.getTaskQueue(taskchecker);
+        for(int i = 0; i < 4; i++) {
+            assertEquals(taskList.get(i), taskQueue.poll());
+        }
+
+        sp.shutDown();
+        Thread.sleep(30);
+
+        assertEquals("ShuttleProcessor shutting down.\n", out.toString());
+    }
+
+    @Test
+    @DisplayName("Should wait until current task has finished evaluating when shut down")
```

Comment:
This test does not seem to work as expected.

* the task should get added via the `addTask()` method since it could contain some essential functionality

* the `ShuttleProcessor` should not be alive after shutting down. It is not expected that the `ShuttleProcessor` waits for the `TaskChecker` in any way.

If the `TaskChecker` completed its job and `ShutDown()` gets called the thread should die:
![image](https://user-images.githubusercontent.com/115566592/215214081-9d3dd101-9f5d-4d3e-bcdc-d5ada6913b52.png)

Reply:
again, I cant really add it via addTask because that breaks some implementations

"the ShuttleProcessor should not be alive after shutting down. It is not expected that the ShuttleProcessor waits for the TaskChecker in any way." 
I think it should stay alive until the task has finished processing. Thats kind of the point of the test. It doesnt wait for taskchecker, the evaluation method just takes 50ms. If you stop the shuttleprocessor before it has finished evaluating, you will skip the part of the code that adds it to task checker (Unless you add it to taskchecker first and then evaluate it hich doesn't make much sense imo)